source/sdk/lang/Exception.ooc

```diff
@@ -1,397 +1,395 @@
 
-    DebugBreak: extern func
-    RaiseException: extern func (ULong, ULong, ULong, Pointer)
-    IsDebuggerPresent: extern func -> Pointer
+	DebugBreak: extern func
+	RaiseException: extern func (ULong, ULong, ULong, Pointer)
+	IsDebuggerPresent: extern func -> Pointer
 }
 
+assert: extern func (Bool)
+
 JmpBuf: cover from jmp_buf {
-    setJmp: extern(setjmp) func -> Int
-    longJmp: extern(longjmp) func (value: Int)
+	setJmp: extern (setjmp) func -> Int
+	longJmp: extern (longjmp) func (value: Int)
 }
 
 _StackFrame: cover {
-    buf: JmpBuf
+	buf: JmpBuf
 }
 
 StackFrame: cover from _StackFrame* {
-    new: static func -> This {
-        gc_malloc(_StackFrame size)
-    }
-}
-
-exceptionStack := ThreadLocal<Stack<StackFrame>> new()
-
-_exception := ThreadLocal<Exception> new()
-_EXCEPTION: Int = 1
-
-stackFramesToDelete := Stack<StackFrame> new()
-
-_pushStackFrame: inline func -> StackFrame {
-  _cleanupStackFrames()
-    stack: Stack<StackFrame>
-    if(!exceptionStack hasValue?()) {
-        stack = Stack<StackFrame> new()
-        exceptionStack set(stack)
-    } else {
-        stack = exceptionStack get()
-    }
-    buf := StackFrame new()
-    stack push(buf)
-    exceptionStack set(stack)
-    buf
-}
-
-_markStackFrame: inline func (frame: StackFrame) {
-  stackFramesToDelete push(frame)
-}
-
-_cleanupStackFrames: inline func {
-  while (!stackFramesToDelete isEmpty) {
-    frame := stackFramesToDelete pop() as StackFrame
-    if (frame)
-      gc_free(frame as Void*)
-  }
-}
-
-_setException: inline func (e: Exception) {
-    _exception set(e)
-}
-
-_getException: inline func -> Exception {
-    _exception get()
-}
-
-_popStackFrame: inline func {
-    frame := exceptionStack get() as Stack<StackFrame> pop() as StackFrame
-    if (frame)
-      gc_free(frame as Void*)
+	new: static func -> This {
+		gc_malloc(_StackFrame size)
+	}
+	free: func {
+		gc_free(this as Void*)
+	}
 }
 
-_takeStackFrame: func -> StackFrame {
-  exceptionStack get() as Stack<StackFrame> pop() as StackFrame
+ExceptionContext: class {
+	_exceptionJumpLabel: Int = 1
+	_localFrames := ThreadLocal<Stack<StackFrame>> new()
+	_currentException := ThreadLocal<Exception> new()
+	_stackFramesToDelete := Stack<StackFrame> new()
+	init: func
+	pushStackFrame: func -> StackFrame {
+		stack: Stack<StackFrame>
+		if (!this _localFrames hasValue?())
+			stack = Stack<StackFrame> new()
+		else
+			stack = this _localFrames get()
+		buf := StackFrame new()
+		stack push(buf)
+		this _localFrames set(stack)
+		buf
+	}
+
+	cleanupStackFrames: func {
+		while (!this _stackFramesToDelete isEmpty) {
+			frame := this _stackFramesToDelete pop() as StackFrame
+			if (frame)
+				frame free()
+		}
+	}
+
+	setException: func (exception: Exception) {
+		this _currentException set(exception)
+	}
+
+	getException: func -> Exception {
+		this _currentException get()
+	}
+
+	popStackFrame: func {
+		frame := this _localFrames get() as Stack<StackFrame> pop() as StackFrame
```

Comment:
I think these explicit casts are a remnant of old versions of rock. Since we're in the business of cleaning up, see if you can remove them.